File: java/src/main/java/com/kcl/KclOption.java
package com.kcl;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import com.kcl.api.Spec.Argument;
import com.kcl.api.Spec.ExecProgramArgs;
import com.kcl.api.Spec.ExecProgramArgs.Builder;
import com.kcl.api.Spec.ExternalPkg;

/**
 * Functional option accepted by the {@link Kcl} facade entry points.
 *
 * <p>
 * Mirrors kcl-go's {@code Option}: each option contributes a slice of state
 * that is merged into a shared {@link KclOptionBag} before the
 * {@link ExecProgramArgs} proto is materialised. Users build options with the
 * static {@code with*} factories on {@link Kcl}; the {@link #apply} method is
 * only callable from within the {@code com.kcl} package because its parameter
 * type is package-private.
 */
@FunctionalInterface
public interface KclOption {
    void apply(KclOptionBag bag);
}

/**
 * Mutable bag of merged option state, mirroring the union of fields the
 * kcl-go {@code Option} struct can populate. Internal to the facade; users
 * only ever see {@link KclOption} instances.
 */
final class KclOptionBag {
    String workDir;
    final List<String> kFilenameList = new ArrayList<String>();
    final List<String> kCodeList = new ArrayList<String>();
    final List<Argument> args = new ArrayList<Argument>();
    final List<String> overrides = new ArrayList<String>();
    final List<String> selectors = new ArrayList<String>();
    final List<ExternalPkg> externalPkgs = new ArrayList<ExternalPkg>();
    String settingsPath;
    Boolean disableNone;
    Boolean sortKeys;
    Boolean showHidden;
    Boolean includeSchemaTypePath;
    Boolean strictRangeCheck;
    Integer verbose;
    Integer debug;

    /**
     * Apply each option in order, mirroring kcl-go's {@code Option.Merge}.
     */
    void applyAll(List<KclOption> options) {
        for (KclOption option : options) {
            option.apply(this);
        }
    }

    /**
     * Materialise the bag into an {@link ExecProgramArgs} proto.
     *
     * <p>
     * When a settings file was recorded via
     * {@code Kcl.withSettings(String)}, it is parsed first and provides the
     * base set of fields; anything the caller populated explicitly through
     * other options is merged on top with kcl-go's {@code Merge} semantics:
     * repeated fields append, non-empty scalars overwrite (last wins), and
     * plain boolean fields only propagate when {@code true} (the zero value
     * means "not set").
     */
    ExecProgramArgs toExecProgramArgs() {
        Builder builder = ExecProgramArgs.newBuilder();
        String baseWorkDir = (workDir == null || workDir.isEmpty()) ? "." : workDir;
        if (settingsPath != null && !settingsPath.isEmpty()) {
            KclSettingsFile.load(settingsPath).applyTo(builder, baseWorkDir);
        }
        if (workDir != null && !workDir.isEmpty()) {
            builder.setWorkDir(workDir);
        }
        builder.addAllKFilenameList(kFilenameList);
        builder.addAllKCodeList(kCodeList);
        builder.addAllArgs(args);
        builder.addAllOverrides(overrides);
        builder.addAllPathSelector(selectors);
        builder.addAllExternalPkgs(externalPkgs);
        if (Boolean.TRUE.equals(disableNone)) {
            builder.setDisableNone(true);
        }
        if (Boolean.TRUE.equals(sortKeys)) {
            builder.setSortKeys(true);
        }
        if (Boolean.TRUE.equals(showHidden)) {
            builder.setShowHidden(true);
        }
        if (Boolean.TRUE.equals(includeSchemaTypePath)) {
            builder.setIncludeSchemaTypePath(true);
        }
        if (Boolean.TRUE.equals(strictRangeCheck)) {
            builder.setStrictRangeCheck(true);
        }
        if (verbose != null && verbose > 0) {
            builder.setVerbose(verbose);
        }
        if (debug != null && debug != 0) {
            builder.setDebug(debug);
        }
        return builder.build();
    }

    static List<String> strings(String... values) {
        return new ArrayList<String>(Arrays.asList(values));
    }
}

File: kotlin/src/main/kotlin/com/kcl/KclOption.kt
package com.kcl

import com.kcl.api.Spec
import com.kcl.api.Spec.ExecProgramArgs

/**
 * Functional option accepted by the [Kcl] facade entry points, mirroring
 * kcl-go's `Option`. Each option contributes a slice of state that is merged
 * into a shared [KclOptionBag] before the [ExecProgramArgs] proto is
 * materialised. Build options with the `with*` factories on [Kcl], or pass a
 * trailing configuration lambda: `Kcl.run(code) { overrides += "a=1" }`.
 */
fun interface KclOption {
    fun apply(bag: KclOptionBag)
}

/**
 * Mutable bag of merged option state, mirroring the union of fields the
 * kcl-go `Option` struct can populate. Constructor is internal; users
 * configure it through option factories or the entry-point DSL lambdas.
 */
class KclOptionBag internal constructor() {
    var workDir: String? = null
    val kFilenames: MutableList<String> = mutableListOf()
    val kCodes: MutableList<String> = mutableListOf()
    val args: MutableList<Spec.Argument> = mutableListOf()
    val overrides: MutableList<String> = mutableListOf()
    val selectors: MutableList<String> = mutableListOf()
    val externalPkgs: MutableList<Spec.ExternalPkg> = mutableListOf()
    var settingsPath: String? = null
    var disableNone: Boolean? = null
    var sortKeys: Boolean? = null
    var showHidden: Boolean? = null
    var includeSchemaTypePath: Boolean? = null
    var strictRangeCheck: Boolean? = null
    var verbose: Int? = null
    var debug: Int? = null

    /**
     * Materialise the bag into an [ExecProgramArgs] proto. When a settings
     * file was recorded it is parsed first and provides the base set of
     * fields; explicitly populated fields are merged on top with kcl-go's
     * `Merge` semantics: repeated fields append, non-empty scalars overwrite
     * (last wins), and plain boolean fields only propagate when `true`.
     */
    internal fun toExecProgramArgs(): ExecProgramArgs {
        val builder = ExecProgramArgs.newBuilder()
        val baseWorkDir = workDir?.takeUnless { it.isEmpty() } ?: "."
        settingsPath?.takeUnless { it.isEmpty() }?.let { path ->
            KclSettingsFile.load(path).applyTo(builder, baseWorkDir)
        }
        workDir?.takeUnless { it.isEmpty() }?.let { builder.workDir = it }
        builder.addAllKFilenameList(kFilenames)
        builder.addAllKCodeList(kCodes)
        builder.addAllArgs(args)
        builder.addAllOverrides(overrides)
        builder.addAllPathSelector(selectors)
        builder.addAllExternalPkgs(externalPkgs)
        if (disableNone == true) builder.disableNone = true
        if (sortKeys == true) builder.sortKeys = true
        if (showHidden == true) builder.showHidden = true
        if (includeSchemaTypePath == true) builder.includeSchemaTypePath = true
        if (strictRangeCheck == true) builder.strictRangeCheck = true
        verbose?.takeIf { it > 0 }?.let { builder.verbose = it }
        debug?.takeIf { it != 0 }?.let { builder.debug = it }
        return builder.build()
    }
}
